### DISCUSSION TITLE
Cannot call async C# function in python

### DISCUSSION THREAD

**Discussion by demirarikan** [pythonnet/pythonnet#2289] [Q&A]
### Environment

-   Pythonnet version: 3.0.3
-   Python version: 3.11.5
-   Operating System: Windows 10
-   .NET Runtime: 4.7.2

### Details

-  Is it possible to call an async C# function in python also asynchronously? 
For example I have this function in C#:
```
 public async Task<string> PerformAsyncOperationAsync(int seconds)
        {
            Console.WriteLine($"Async operation started. Waiting for {seconds} seconds.");
            await Task.Delay(TimeSpan.FromSeconds(seconds));
            Console.WriteLine($"Async operation completed after {seconds} seconds.");

            return $"Operation result after {seconds} seconds";
        }
```
and I want to ideally call it like this in python:
```
import clr
import asyncio

clr.AddReference(r'C:\Users\demir.arikan\Desktop\iLabs2\dllTest\dlls\AsyncTest.dll')

from AsyncLibrary import AsyncClass

async def main():
    async_instance = AsyncClass()

    result = await async_instance.PerformAsyncOperationAsync(3)

    print(result)
    asyncio.run(main())
```
But this gives me the error that I cannot use await here. When the await is removed the function just returns a Task object and does not execute properly. 

I saw in a previous thread that async to async wasn't implemented yet and but this was 5 years ago. Are there any updates about this? Is it possible to call the async C# functions without blocking the threads (.Result or wait)?

**Comment by Otamam** [comment]
`
import clr
import asyncio

clr.AddReference(r'C:\Users\demir.arikan\Desktop\iLabs2\dllTest\dlls\AsyncTest.dll')

from AsyncLibrary import AsyncClass

def main(number):

    async def foo_inner_async():
        result = Optimizer.OptimizeAsync(number)
        result.Wait()
        return result.Result

    loop = asyncio.get_event_loop()
    res = loop.run_until_complete(foo_inner_async())
    return res
`
